### DISCUSSION TITLE
module 'sqlalchemy.sql.schema' has no attribute '_schema_getter

### DISCUSSION THREAD

**Discussion by kershnerd** [python-gino/gino#765] [Q&A]
This is so weird and I am out of ideas.  I had code that has been working for a long time.  But how I get this error and based on the logs, nothing has changed.

Here is the exact error:

```
analytics_1  | Traceback (most recent call last):
analytics_1  |   File "manage.py", line 6, in <module>
analytics_1  |     app = create_app()
analytics_1  |   File "/usr/src/app/services/analytics/server.py", line 50, in create_app
analytics_1  |     from services.common.models import db
analytics_1  |   File "/usr/src/app/services/common/models/__init__.py", line 5, in <module>
analytics_1  |     from services.common.models.base import db
analytics_1  |   File "/usr/src/app/services/common/models/base.py", line 1, in <module>
analytics_1  |     from gino.ext.sanic import Gino
analytics_1  |   File "/usr/local/lib/python3.8/site-packages/gino/__init__.py", line 2, in <module>
analytics_1  |     from .engine import GinoEngine, GinoConnection  # NOQA
analytics_1  |   File "/usr/local/lib/python3.8/site-packages/gino/engine.py", line 177, in <module>
analytics_1  |     class GinoConnection:
analytics_1  |   File "/usr/local/lib/python3.8/site-packages/gino/engine.py", line 207, in GinoConnection
analytics_1  |     schema_for_object = schema._schema_getter(None)
analytics_1  | AttributeError: module 'sqlalchemy.sql.schema' has no attribute '_schema_getter'
```

I'm in a python 3.8 docker image and here is my requirements.txt file:
```
gunicorn==19.9.0
sanic==19.6.3
sanic-cors==0.9.9.post3
gino==0.8.3
click==7.0
requests==2.22.0
python-jose-cryptodome==1.3.2
six==1.12.0
```

Any ideas on where I can even start to track this error down?

**Comment by mikekeda** [comment]
It's a new version of sqlalchemy==1.4.0, right?
I think gino isn't adapted to the new sqlalchemy version yet (1.4.0 was released yesterday)

**Reply by kershnerd** [reply]
I don't explicitly install sqlalchemy.  I just  install the requirements.txt:

```
gunicorn==19.9.0
sanic==19.6.3
sanic-cors==0.9.9.post3
gino==0.8.3
click==7.0
requests==2.22.0
python-jose-cryptodome==1.3.2
six==1.12.0
```

**Reply by mikekeda** [reply]
gino requires sqlalchemy,
run
`pip show sqlalchemy`
to check sqlalchemy version

**Comment by jrversteegh** [comment]
Any progress on this? Still not working after half a year? Gino 1.0.1, SqlAlchemy 1.4.27, same issue still.

**Reply by wwwjfy** [reply]
Gino at this moment is not compatible with SQLAlchemy 1.4. The support is not being worked on.
can downgrade to 1.3 or not install SQLAlchemy directly or indirectly

https://github.com/python-gino/gino/blob/master/pyproject.toml#L27

**Reply by jrversteegh** [reply]
OK, would it make sense to improve the error message and state this explicitly? I'd be willing to create a pull request for that..

**Reply by wwwjfy** [reply]
I think what we can do is during initialisation, do sanity check on sqlalchemy (and other dependencies as well) version, but not specifically this error. But honestly, I think this should be handled by dependency management tool instead of each library.
poetry would block it if there were conflicts, like it requires sqlalchemy >= 1.4 and < 1.4 from different dependencies.

**Reply by jrversteegh** [reply]
Poetry may handle this properly, but pip does not. I agree that a package manager _should_ handle this, but it doesn't work apparently and probably won't in the future. Adding that to a case where you specifically not support the latest version of package and have no intention of doing so, it seems helpful to alert users to this rather than have them waste 20 minutes to figure this rather trivial issue out through googling like I did :p.

**Reply by wwwjfy** [reply]
`pip check` can verify if there is any conflict.
I don't know if we already have a library to check the discrepancies between installed libraries and specified dependencies in this library. (There probably is)

But I'm a bit reluctant to do so because I personally do not want to enforce users on this, as they may have valid cases to have libraries not satisfying what is specified (very likely the version range is made too narrow) and they can't proceed because of this.
Maybe we can print warnings at the initialisation stage, but I doubt users will look at the warnings even when they meet such issues. But I'm fine to review and merge if that's the proposal. Up to discussion if there are other ideas.

**Reply by SpudInNZ** [reply]
If gino is not going to support current versions of sqlalchemy, our organisation will have to stop using it.
